src/ray/gcs/store_client/redis_store_client.h

```diff
@@ -0,0 +1,135 @@
+#ifndef RAY_GCS_STORE_CLIENT_REDIS_STORE_CLIENT_H
+#define RAY_GCS_STORE_CLIENT_REDIS_STORE_CLIENT_H
+
+#include <memory>
+#include <unordered_set>
+#include "ray/gcs/redis_client.h"
+#include "ray/gcs/redis_context.h"
+#include "ray/gcs/store_client/store_client.h"
+
+namespace ray {
+
+namespace gcs {
+
+class RedisStoreClient : public StoreClient {
+ public:
+  RedisStoreClient(const StoreClientOptions &options);
+
+  virtual ~RedisStoreClient();
+
+  Status Connect(std::shared_ptr<IOServicePool> io_service_pool) override;
+
+  void Disconnect() override;
+
+  Status AsyncPut(const std::string &table_name, const std::string &key,
+                  const std::string &value, const StatusCallback &callback) override;
+
+  Status AsyncPut(const std::string &table_name, const std::string &key,
+                  const std::string &index, const std::string &value,
+                  const StatusCallback &callback) override;
+
+  Status AsyncGet(const std::string &table_name, const std::string &key,
+                  const OptionalItemCallback<std::string> &callback) override;
+
+  Status AsyncGetByIndex(const std::string &table_name, const std::string &index,
+                         const MultiItemCallback<std::string> &callback) override;
+
+  Status AsyncGetAll(const std::string &table_name,
+                     const ScanCallback<std::string, std::string> &callback) override;
+
+  Status AsyncDelete(const std::string &table_name, const std::string &key,
+                     const StatusCallback &callback) override;
+
+  Status AsyncDeleteByIndex(const std::string &table_name, const std::string &index,
+                            const StatusCallback &callback) override;
+
+ private:
+  Status DoPut(const std::string &key, const std::string &value,
+               const StatusCallback &callback);
+
+  std::shared_ptr<RedisClient> redis_client_;
+};
+
+/// \class RedisRangeOpExecutor
+/// This class is used for three kind of operations:
+/// 1. Get All Data with the same prefix from Redis (GetAll).
+/// 2. Get All Data Key with the same index from Redis (GetByIndex).
+/// 3. Delete Index and Data Key with the same index from Redis (DeleteByIndex).
+/// TODO(micafan) Consider encapsulating three different classes:
+/// Scanner, MultiReader, BatchDeleter.
+class RedisRangeOpExecutor : public std::enable_shared_from_this<RedisRangeOpExecutor> {
+ public:
+  RedisRangeOpExecutor(std::shared_ptr<RedisClient> redis_client,
+                       const std::string &table_name, const std::string &index,
+                       const MultiItemCallback<std::string> &get_by_index_callback);
+
+  RedisRangeOpExecutor(std::shared_ptr<RedisClient> redis_client,
```

Comment:
It's hard to tell the difference between the constructors. Think about this, If I write `RedisRangeOpExecutor(my_table, my_index, my_callback)`, do you know which range operation I'm trying to run?

One solution is to divide this class into a base class and three inherited classes, and give the inherited classes distinguishable names. This may involve some refactoring. e.g. `RedisGetAllDataWithPrefixExecutor(my_table, my_index, my_callback)`.

Another solution is to use factory functions. e.g. `RedisRangeOpExecutor::NewGetAllDataWithPrefixExecutor(my_table, my_index, my_callback)`.

Reply:
As I comment in head file of RedisRangeOpExecutor, I want to add  three new class: Scanner, MultiReader, BatchDeleter. To make the operator class easy extension. And the range operation will be defined by enum: KScanDataTable, KDeleteByIndex, KScanByIndex....

But the workload will be larger than it is now and it will take more time. And `lingxun` is  required `RedisStoreClient` for a long time, so I write this version to unblock him.